neuralmagic/sparseml
# SparseML Compression Pt 3: SparseAutoModel interface & inferring params

This final PR for SparseML sparsity compression adds compression support to `save_pretrained` and `from_pretrained`. It also adds support for inferring global sparsity and sparsity structure params from the model. See the [corresponding internal docs PR](https://github.com/neuralmagic/internal-docs/pull/42) for design details, but there have been some minor changes that are reflected in [README.md](https://github.com/neuralmagic/sparseml/blob/a2c8ff20bd1099d3e335e0ccdee4f17bb5595f22/src/sparseml/transformers/compression/README.md)


### Callouts
* Overriding `model.save_pretrained()` is tricky, because the model is initialized from `SparseAutoModelForCausalLM.from_pretrained`, but the returned model is a child of `PreTrainedModel`(for instance `LlamaForCausalLM` inherits from `LlamaPreTrainedModel` inherits from `PreTrainedModel`) To override `PreTrainedModel.save_pretrained` we need to do a bit of "class instance surgery". See `compression_save.py` for implementation details
* Ended up not storing the sparsity config in the safetensors metadata, storing it in the HF config seemed sufficient and I felt it was going to complicate the code without much added gain
* How we're inferring the sparsity structure is definitely hacky. We determine it by looking at the applied modifiers and checking for a mask_structure attribute. I think this is sufficient for now, but open to other ideas here

### Example 
```python
from sparseml.transformers import SparseAutoModelForCausalLM
from sparseml.utils.pytorch.utils import measure_cuda_memory
import torch

MODEL_PATH = "zoo:llama2-7b-open_platypus_orca_llama2_pretrain-pruned60"
OUTPUT_PATH = "./test_compress_output"
RECIPE = "zoo:llama2-7b-open_platypus_orca_llama2_pretrain-pruned60"

torch.cuda.set_device(0)
with measure_cuda_memory() as m:
    model = SparseAutoModelForCausalLM.from_pretrained(MODEL_PATH, device_map="cuda:0")
print(f"Load dense model peak GPU {m.overall_peak_memory / float(2**30):.4f} GB")

sparsity_config = getattr(model,"sparsity_config", None)
print(f"Sparsity config before compression: {sparsity_config}")
with measure_cuda_memory() as m:
    model.save_compressed(OUTPUT_PATH)
print(f"Save compressed model peak GPU {m.overall_peak_memory / float(2**30):.4f} GB")

torch.cuda.set_device(1)
with measure_cuda_memory() as m:
    model_again = SparseAutoModelForCausalLM.from_pretrained(
        OUTPUT_PATH, device_map="cuda:1"
    )
print(f"Load compressed model peak GPU {m.overall_peak_memory / float(2**30):.4f} GB")
sparsity_config = getattr(model,"sparsity_config", None)
print(f"Sparsity config after compression: {sparsity_config}")

og_state_dict = model.state_dict()
reconstructed_state_dict = model_again.state_dict()
assert(len(og_state_dict) == len(reconstructed_state_dict))
for key in og_state_dict.keys():
    dense_tensor = og_state_dict[key]
    reconstructed_tensor = reconstructed_state_dict[key]
    assert torch.equal(dense_tensor.cpu(), reconstructed_tensor.cpu())
```

Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:03<00:00,  1.55it/s]
Load dense model peak GPU 25.2276 GB
Sparsity config before compression: None
Compressing model: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 291/291 [01:22<00:00,  3.51it/s]
Save compressed model peak GPU 26.3272 GB
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  7.87it/s]
Decompressing model: 291it [00:23, 12.31it/s]
Load compressed model peak GPU 25.7159 GB
Sparsity config after compression: format='sparse_bitmask' global_sparsity=57.66354988216862 sparsity_structure='unstructured'

### Testing
Unit tests are included for `from_pretrained`, `save_pretrained` and `save_compressed`. In addition I manually tested the following scenarios. See the README for instructions on turning on compression during finetuning/oneshot
* load dense -> apply SparseGPT -> save compressed -> reload compressed
* FSDP: load dense -> apply SparseGPT -> save compressed -> reload compressed
* load sparse -> sparse finetuning -> save compressed -> reload compressed
* FSDP: load sparse -> sparse finetuning -> save compressed -> reload compressed
